numpy/random/mtrand/mtrand.pyx

```diff
@@ -5064,7 +5064,8 @@ cdef class RandomState:
             x_ptr = <char*><size_t>x.ctypes.data
             stride = x.strides[0]
             itemsize = x.dtype.itemsize
-            buf = np.empty_like(x[0])  # GC'd at function exit
+            # Ensure that buf has the same dtype as x
+            buf = np.empty_like(x[:1])  # GC'd at function exit
```

Comment:
OK, I think the object cases were probably already all fine (though I am not 100% sure, but oh wlel...)